Reference disk usage metric + refactoring [BT-20]

grsterin opened the pull request.

kshakir reviewed (commented):
Main code LGTM so far. Due to parallel CI failures I can't tell if this was integration tested in centaur or not, but the changes do not appear to be unit tested by the `sbt` suite, especially `sendNumberOfReferenceFilesUsedGauge`.

Also, are there any docs on how to use this? For example, will this new metric automatically appear alongside the others in the metrics explorer on the [Cromwell Dashboard](https://console.cloud.google.com/monitoring/dashboards/builder/12440147339961148290?project=broad-dsde-prod), does something need to be done to further wire this in?

grsterin commented on supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala:
```diff
@@ -623,6 +637,20 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
+      case (true, Some(referenceFileToDiskImageMapping)) =>
+        val referenceInputsNumber = getReferenceInputsToMountedPathMappings(referenceFileToDiskImageMapping, inputs).size
+        sendGauge(
+          NonEmptyList.of("referencefiles", "used", "number"),
```
You mean incorporating file name into the metric "name"? I think the number of reference files is huge and in Grafana we'd need to manually add each metric to the dashboard to make it visible

mcovarr replied:
Hmm I didn't think those needed to be created manually but I am not a Grafana expert. We certainly don't create new metrics manually when we spin up new k8s pods.

mcovarr replied:
There are currently 30 reference files set to go out, 24 for HG 38 and 6 for HG 19.

grsterin replied:
Yeah, metrics have to be manually added to the dashboard to be visible. But now I remember that Grafana supports wildcards, so we can add metrics for all files as a single query

grsterin commented on supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala:
```diff
@@ -623,6 +637,20 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
+          InstrumentationPrefixes.JobPrefix
+        )
+      case _ =>
+        // reference disks feature is either not configured in Cromwell or disabled in workflow options
```
I'm not sure we currently write in the log if any of PAPI configuration attributes are specified or not.

mcovarr replied:
any chance of sneaking a one-liner like that into this PR or should we make a separate ticket?

kshakir reviewed (approved):
For folks that come across this PR in the future, we did NOT make time today to discuss where DSP metrics are supposed to be sent... so I believe this PR is still generating metrics aimed at the old dashboard and not what folks have been referring to as the "post-kubernetes" dashboard 🤷🦶🏈

mcovarr commented on supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/PipelinesApiAsyncBackendJobExecutionActor.scala:
```diff
@@ -176,52 +174,37 @@ class PipelinesApiAsyncBackendJobExecutionActor(standardParams: StandardAsyncExe
-            case (input, absolutePathOnRefDisk) =>
-              s"mkdir -p ${input.containerPath.parent.pathAsString} && ln -s $absolutePathOnRefDisk ${input.containerPath.pathAsString}"
-          }
+    val referenceFilesLocalizationScript = {
```
It seems surprising the lines below are marked as not being tested, unless Centaur coverage is not being included for some reason...

grsterin commented:
Updated to publish counter metrics on per-reference-file basis (incremented until Cromwell restart). Also added logging for reference disks feature configuration step.

mcovarr commented on supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActorSpec.scala:
```diff
@@ -127,12 +131,13 @@ class PipelinesApiAsyncBackendJobExecutionActorSpec extends TestKitSuite
         DefaultStandardAsyncExecutionActorParams(
           jobIdKey = PipelinesApiAsyncBackendJobExecutionActor.JesOperationIdKey,
-          serviceRegistryActor = kvService,
+          serviceRegistryActor = serviceRegistryActor,
```
oops

grsterin commented on supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala:
```diff
@@ -623,6 +648,17 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
+    referenceInputFilesOpt match {
+      case Some(referenceInputFiles) =>
+        referenceInputFiles.foreach { referenceInputFile =>
+          increment(NonEmptyList.of("referencefiles", referenceInputFile.relativeHostPath.pathAsString))
```
Yes, it's going to be the same as cloud path, but with `gs://` part dropped. Like `gcp-public-data--broad-references/hg19/v0/README`
